### PR TITLE
upsert treatments based on created_at and eventType

### PR DESCRIPTION
**Pull request by jasoncalabrese**
also dedupe and process data within a posted array
more treatment api tests

### CODE REVIEW

File: lib/treatments.js [hunk 1]
```diff
@@ -24,14 +59,15 @@ function storage (env, ctx) {
           pbTreat.notes = obj.notes;
         }
 
-        api( ).insert(pbTreat, function() {
-          //nothing to do here
+        query.created_at = pbTreat.created_at;
+        api( ).update(query, pbTreat, {upsert: true}, function complete (err) {
```
Comment (jasoncalabrese) [inline lib/treatments.js]: Codacy Issue found: [Function 'complete' is already declared in line 49.](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1722678008/issues/source?bid=2452379&fileBranchId=3051501#l63)

File: lib/treatments.js [hunk 2]
```diff
@@ -1,18 +1,53 @@
...
+          errs.push(err);
+          if (totalReturn === objOrArray.length) {
+            errs = _.compact(errs);
+            done(errs.length > 0 ? errs : null, allDocs);
+          }
+        });
+      });
+    } else {
+      upsert(objOrArray, function upserted (err, docs) {
+        done(err, docs)
```
Comment (jasoncalabrese) [inline lib/treatments.js]: Codacy Issue found: [Missing semicolon.](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1722678008/issues/source?bid=2452379&fileBranchId=3051501#l33)

Comment (jasoncalabrese) [inline lib/treatments.js]: Codacy Issue found: [Missing semicolon.](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1722711126/issues/source?bid=2452379&fileBranchId=3051501#l33)

**Comment by mgranberry** [comment]
It appears to work when I use alrightypump-commandline's uploader.  I had to stop using client-created IDs in order to get it to work.

**Comment by jasoncalabrese** [comment]
We can add support for client-created IDs if you, but might be better to use a different field.

I'm planning to test this later today/tonight

**Comment by mgranberry** [comment]
I was only using it as a poor man's deduplication method before I got
around to tracking what records have been sent.

**Pull request merged**